### PR TITLE
[PHP] Add ParamAndReturnScalarTypehintsRector

### LINKED ISSUE #794: PHP 7.1 typehints to code, same as properties typehints
### Related

- https://github.com/nikic/TypeUtil
- https://github.com/dunglas/phpdoc-to-typehint
- https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3258

_Originally posted by @TomasVotruba in https://github.com/rectorphp/rector/issues/638#issuecomment-441487799_

### Cases to Cover

```php
<?php

interface Sniff
{
    /**
     * @param int $position
     */
    public function proces(File $file, $position);
}
```

Refactored code:

```php
<?php

final class MySniff implements Sniff
{
    /**
     * @param int $position
     */
    public function proces(File $file, $position)
    {
    }
}
```

### PR DESCRIPTION
**Pull request by TomasVotruba**
Closes #794 

### Checklist

- [x] keep external scope of class implemented interface method
- [x] keep external scope of parent class extended class method
- [x] change local scope of class implemented interface method
- [x] change local scope of parent class extended class method

### Adopt tests

- [x] https://github.com/nikic/TypeUtil/
- [x] https://github.com/dunglas/phpdoc-to-typehint
- https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3640
   - [x] @return https://github.com/Slamdunk/PHP-CS-Fixer/blob/d7a409c10d0e21bc847efb26552aa65bb3c61547/tests/Fixer/FunctionNotation/PhpdocToReturnTypeFixerTest.php
- https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4056/
    - [x] @param https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/4a47b6df0bf718b49269fa9920b3723d802332dc/tests/Fixer/FunctionNotation/PhpdocToParamTypeFixerTest.php

### Resolve Issues

- [x] https://github.com/nikic/TypeUtil/issues
- [x] https://github.com/dunglas/phpdoc-to-typehint/issues

### CODE REVIEW

File: packages/Php/src/Rector/ClassMethod/ParamAndReturnScalarTypehintsRector.php [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+                    return false;
+                }
+
+                // if not, look for it's parent parent - @todo recursion
+            }
+
+            if ($this->broker->hasClass($parentClassName)) {
+                $parentReflectionClass = new ReflectionClass($parentClassName);
+                if ($parentReflectionClass->hasMethod($methodName)) {
+                    // @todo validate type is conflicting
```
Comment (TomasVotruba) [inline packages/Php/src/Rector/ClassMethod/ParamAndReturnScalarTypehintsRector.php]: Maybe `method_exists`

**Comment by bendavies** [comment]
awesome!

**Comment by TomasVotruba** [comment]
I thought I explaind you difference between tokens and AST during the call.

These 3 should sum it up:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4056#issuecomment-442264393

Any feedback on the coe?

**Comment by TomasVotruba** [comment]
Let's get feedback on this :) merging

**Comment by bendavies** [comment]
Think this is missing from the docs?

**Comment by bendavies** [comment]
Ah no, it is.

**Comment by bendavies** [comment]
quick test.
This doesn't handle `@return $this` well, where the user probably meant `@return self`

**Comment by TomasVotruba** [comment]
@bendavies  `$this` fix ready https://github.com/rectorphp/rector/pull/820/files#diff-1317f3d44882779fd2643bc3488741f2R159

**Comment by bendavies** [comment]
hey @TomasVotruba what are these rectors called now? i could not find them - i guess they have been renamed
I was going from your blog post: https://www.tomasvotruba.cz/blog/2018/12/10/rocket-science-behind-migration-of-docblock-types-to-php-typehints/

**Comment by bendavies** [comment]
i got them

Rector\Php\Rector\FunctionLike\ParamTypeDeclarationRector
Rector\Php\Rector\FunctionLike\ReturnTypeDeclarationRector

**Comment by TomasVotruba** [comment]
I got to it while coding `void` types, done :heavy_check_mark: 

https://github.com/TomasVotruba/tomasvotruba.cz/commit/6304ea48524f082de9f1247ec4cb666162b9454d

**Pull request merged**